master/mongodb_assigner.cpp

```diff
@@ -0,0 +1,164 @@
+}
+
+void MongoSplitAssigner::master_mongodb_req_end_handler() {
+    auto& master = Master::get_instance();
+    auto* resp_socket = master.get_socket();
+    BinStream stream = zmq_recv_binstream(resp_socket);
+    MongoDBSplit split;
+    stream >> split;
+    recieve_end(split);
+
+    stream.clear();
+    zmq_sendmore_string(resp_socket, master.get_cur_client());
+    zmq_sendmore_dummy(resp_socket);
+    zmq_send_binstream(resp_socket, stream);
+    base::log_msg("master => end@" + split.get_ns() + " " + split.get_min() + ":" + split.get_max());
+}
+
+void MongoSplitAssigner::master_setup_handler() {
+}
+
+MongoSplitAssigner::~MongoSplitAssigner() {
+    shards_map_.clear();
+    splits_.clear();
+    splits_end_.clear();
+}
+
+void MongoSplitAssigner::set_auth(const std::string& username, const std::string& password) {
+    username_ = username;
+    password_ = password;
+    need_auth_ = true;
+}
+
+void MongoSplitAssigner::reset_auth() { need_auth_ = false; }
+
+void MongoSplitAssigner::create_splits() {
+    mongo::DBClientConnection client;
+    client.connect(server_, error_msg_);
+
+    if (need_auth_)
+        client.auth(kConfigDB, username_, password_, error_msg_);
+
+    auto shard_cursor = client.query(kConfigShards);
+    while (shard_cursor->more()) {
+        mongo::BSONObj o = shard_cursor->next();
+        std::string shard_id = o.getStringField("_id");
+        std::string shard_host = o.getStringField("host");
+        shards_map_.insert(std::pair<std::string, std::string>(shard_id, shard_host));
+    }
+
+    mongo::BSONObjBuilder ns_filter;
+    ns_filter.append("ns", ns_);
+    auto chunk_cursor = client.query(kConfigChunks, ns_filter.obj());
+    while (chunk_cursor->more()) {
+        mongo::BSONObj o = chunk_cursor->next();
+
+        MongoDBSplit split;
+        split.set_ns(ns_);
+
+        std::string shard = o.getStringField("shard");
+        std::string shard_uri = shards_map_.find(shard)->second;
+        split.set_input_uri(shard_uri);
+
+        split.set_min(o.getObjectField("min").jsonString());
+        split.set_max(o.getObjectField("max").jsonString());
+
+        split.set_valid(true);
+        splits_.push_back(split);
+    }
+    split_num_ = splits_.size();
+}
+
+MongoDBSplit MongoSplitAssigner::answer(const std::string& server, const std::string& ns) {
+    if (server_ != server || ns_ != ns) {
+        server_ = server;
+        ns_ = ns;
+        create_splits();
+    }
+
+    if (splits_.empty()) {
+        if (end_count_ == split_num_) {
+            end_count_ = 0;
+            splits_.swap(splits_end_);
+            splits_end_.clear();
+        }
+        return MongoDBSplit();
+    }
+
+    MongoDBSplit ret = splits_.back();
+    splits_.pop_back();
+    return ret;
+}
+
+void MongoSplitAssigner::recieve_end(MongoDBSplit& split) {
+    if (!split.is_valid())
+        return;
+
+    end_count_++;
+    splits_end_.push_back(split);
+}
+
+static MongoSplitAssigner mongo_split_assigner;
```

Comment:
Could you put line 160 in line 30? The static variable feels like to be hidden in the end of this file.